### PR TITLE
Virtualization Guest Migration Test

### PR DESCRIPTION
**Pull request by jerrytt**
Vitualization Multi-machine job  :  Guest Migration
We have 2 role : 
 1.The source host , which will migrate to guest to the target host.
 2.The target host, wich will take the guest from source host.

In this test: 
  "Target host" play the parent role.
  "Source host"  play the child role.

we should invoke the child to let the schedule trigger the parent.

Migration step (S stand for source host, T stand for target host) : 
1. both S and T perform the  os installation 
2  both S and T perform the RPM update
3  both S and T switch the hyper viser (option)
4  both S and T setup the virtualization Migration preparation .
5  S setup nfs  ,   T sync the nfs setup .
6  S and T mount share nfs.
7  S install the GUEST
8  S do the Migration

### CODE REVIEW

File: tests/virt_autotest/guest_migration_base.pm [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+package guest_migration_base;
+use base "opensusebasetest";
+use strict;
+use testapi;
+use utils;
+use lockapi;
+use mmapi;
+use Data::Dumper;
+use Exporter qw(import);
```
Comment (okurz) [inline tests/virt_autotest/guest_migration_base.pm]: please only import what you need

File: tests/virt_autotest/guest_migration_base.pm [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    $vm_disk_dir                      = "/var/lib/$hyper_visor/images";
+    $install_script                   = "/usr/share/qa/qa_test_virtualization/installos";
+    $guest_install_prepare_keep_guest = qq(sed -i 's/INSTALL_METHOD$/& -g/' $install_script);
+    $install_script .= " standalone";
+}
+
+$nfs_local_dir = "/tmp/pesudo_mount_server";
+
+sub get_var_from_parent {
+    my $var     = shift;
```
Comment (okurz) [inline tests/virt_autotest/guest_migration_base.pm]: please use style `my ($var) = @_;`

File: tests/virt_autotest/virt_autotest_base.pm [hunk 3]
```diff
@@ -95,7 +95,7 @@ sub execute_script_run {
 
     if ($ret) {
         save_screenshot;
-        $ret =~ s/$pattern//g;
+        $ret =~ s/[\r\n]+$pattern[\r\n]+//g;
```
Comment (okurz) [inline tests/virt_autotest/virt_autotest_base.pm]: That should probably go into its own commit with a good explanation of the reason of the change.

Comment (jerrytt) [inline tests/virt_autotest/virt_autotest_base.pm]: This change just clean up the invisible character cause by verify code.(echo CMD-FINISHED-xxxxx) , and this PR require this to get output inside the worker.
I'm ok that cancel this change , and open another pr before this pr

File: tests/virt_autotest/guest_migration_base.pm (hunk 2 above)
Comment (jerrytt) [inline tests/virt_autotest/guest_migration_base.pm]: sure , thanks for notice

File: tests/virt_autotest/guest_migration_base.pm (hunk 1 above)
Comment (jerrytt) [inline tests/virt_autotest/guest_migration_base.pm]: sure , thanks for notice

File: tests/virt_autotest/virt_autotest_base.pm (hunk 3 above)
Comment (okurz) [inline tests/virt_autotest/virt_autotest_base.pm]: not another PR but another commit in the same PR

**Comment by jerrytt** [comment]
I'll and another separate commit for removing the invisible character

File: tests/virt_autotest/virt_autotest_base.pm (hunk 3 above)
Comment (jerrytt) [inline tests/virt_autotest/virt_autotest_base.pm]: @okurz does the separate commit  fine for this PR?

**Comment by okurz** [comment]
Unfortunately there is a typo in the commit message s/invisiable/invisible/

**Comment by jerrytt** [comment]
@okurz just fix the typo

File: products/sle/main.pm [hunk 4]
```diff
@@ -1078,6 +1078,16 @@ elsif (get_var("VIRT_AUTOTEST")) {
...
+    elsif (get_var("VIRT_PRJ3_GUEST_MIGRATION_SOURCE")) {
+        loadtest "virt_autotest/guest_migration_config_virtualization_env.pm";
+        loadtest "virt_autotest/guest_migration_source_nfs_setup.pm";
+        loadtest "virt_autotest/guest_migration_source_install_guest.pm";
+        loadtest "virt_autotest/guest_migration_source_migrate.pm";
+    }
+    elsif (get_var("VIRT_PRJ3_GUEST_MIGRATION_TARGET")) {
+        loadtest "virt_autotest/guest_migration_config_virtualization_env.pm";
+        loadtest "virt_autotest/guest_migration_target_nfs_setup.pm";
+    }
```
Comment (okurz) [inline products/sle/main.pm]: Please rebase and remove all `.pm` suffixes

File: tests/virt_autotest/guest_migration_base.pm [hunk 5]
```diff
@@ -0,0 +1,93 @@
+# SUSE's openQA tests
+#
+# Copyright © 2012-2016 SUSE LLC
```
Comment (okurz) [inline tests/virt_autotest/guest_migration_base.pm]: Please specify only the year during which the file was changed, I guess 2016 only

File: tests/virt_autotest/guest_migration_config_virtualization_env.pm [hunk 6]
```diff
@@ -0,0 +1,47 @@
+# SUSE's openQA tests
+#
+# Copyright © 2012-2016 SUSE LLC
```
Comment (okurz) [inline tests/virt_autotest/guest_migration_config_virtualization_env.pm]: here, too

File: tests/virt_autotest/guest_migration_source_install_guest.pm [hunk 7]
```diff
@@ -0,0 +1,33 @@
+# SUSE's openQA tests
+#
+# Copyright © 2012-2016 SUSE LLC
```
Comment (okurz) [inline tests/virt_autotest/guest_migration_source_install_guest.pm]: here, too

File: tests/virt_autotest/guest_migration_config_virtualization_env.pm [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+# Summary: virt_autotest: Virtualization multi-machine job : Guest Migration
+# Maintainer: jerry <jtang@suse.com>
+
+use base virt_autotest_base;
+use strict;
+use testapi;
+use guest_migration_base;
+
+sub run() {
+    my $self = shift;
```
Comment (okurz) [inline tests/virt_autotest/guest_migration_config_virtualization_env.pm]: please use array assignment instead of shift operator, i.e. `my ($self) = @_;

File: tests/virt_autotest/guest_migration_config_virtualization_env.pm [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    if ($hyper_visor =~ /xen/) {
+
+        $self->execute_script_run("source /usr/share/qa/virtautolib/lib/virtlib;changeXendConfig", 500);
+
+    }
+    else {
+
+        $self->execute_script_run("source /usr/share/qa/virtautolib/lib/virtlib;changeLibvirtConfig", 500);
+
+    }
```
Comment (okurz) [inline tests/virt_autotest/guest_migration_config_virtualization_env.pm]: delete blank lines

File: tests/virt_autotest/guest_migration_source_migrate.pm [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+# Summary: virt_autotest: Virtualization multi-machine job : Guest Migration
+# Maintainer: jerry <jtang@suse.com>
+
+use base virt_autotest_base;
+use strict;
+use testapi;
+use guest_migration_base;
+
+sub analyzeResult {
+
```
Comment (okurz) [inline tests/virt_autotest/guest_migration_source_migrate.pm]: delete blank line

File: tests/virt_autotest/guest_migration_source_migrate.pm [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+                $result->{$case_name}{status} = "FAILED" if ($case_result =~ /fail/);
+                $result->{$case_name}{time}   = 1;
+            }
+        }
+    }
+    return $result;
+}
+
+sub upload_tar_log() {
+    my $self = shift;
```
Comment (okurz) [inline tests/virt_autotest/guest_migration_source_migrate.pm]: `my ($self) = @_;`

**Comment by okurz** [comment]
thanks for your effort, merging.
